src/dir.cr

```diff
@@ -255,29 +255,147 @@ class Dir
   end
 
   def self.glob(patterns : Enumerable(String))
-    paths = LibC::Glob.new
-    flags = LibC::GlobFlags::BRACE | LibC::GlobFlags::TILDE
-    errfunc = -> (_path : UInt8*, _errno : Int32) { 0 }
+    path = self.working_directory
+    regexes = patterns.map { |pattern| glob2regex(pattern) }
+    if patterns.any? { |ptrn| ptrn.includes? "**" }
+      recursion_depth = Int32::MAX
+    else
+      recursion_depth = patterns.map { |ptrn| ptrn.count(File::SEPARATOR) }.max
+    end
+    scandir(path, "", regexes, 0, recursion_depth) do |path|
+      yield path
+    end
+  end
+  
+  private def self.glob2regex(pattern)
+    if pattern.size == 0 || pattern == File::SEPARATOR
+      raise ArgumentError.new "Empty glob pattern"
+    end
+
+    # characters which are escapable by a backslash in a glob pattern;
+    # Windows paths must have double backslashes:
+    escapable = ['?', '{', '}', '*', ',', '\\']
+    # characters which must be escaped in a PCRE regex:
+    escaped = ['$', '(', ')', '+', '.', '[', '^', '|', '/']
 
-    patterns.each do |pattern|
-      result = LibC.glob(pattern, flags, errfunc, pointerof(paths))
+    regex_pattern = String.build do |str|
+      idx = 0
+      nest = 0
 
-      if result == LibC::GlobErrors::NOSPACE
-        raise GlobError.new "Ran out of memory"
-      elsif result == LibC::GlobErrors::ABORTED
-        raise GlobError.new "Read error"
+      idx = 1 if pattern[0] == File::SEPARATOR
+      if pattern.starts_with? "**#{File::SEPARATOR}"
+        idx = 3
+        str << ".*"
       end
 
-      flags |= LibC::GlobFlags::APPEND
+      while idx < pattern.size
+        if pattern[idx] == '\\'
+          if idx + 1 < pattern.size && escapable.includes? pattern[idx + 1]
+            str << '\\'
+            str << pattern[idx + 1]
+            idx += 2
+            next
+          end
+        elsif pattern[idx] == '*'
+          if idx + 1 < pattern.size && pattern[idx + 1] == '*'
+            str << ".*"
+            idx += 2
+            next
+          else
+            str << "[^\\#{File::SEPARATOR}]*"
+          end
+        elsif escaped.includes? pattern[idx]
+          str << "\\"
+          str << pattern[idx]
+        elsif pattern[idx] == '?'
+          str << "[^\\#{File::SEPARATOR}]"
+        elsif pattern[idx] == '{'
+          str << "(?:"
+          nest += 1
+        elsif pattern[idx] == '}'
+          str << ")"
+          nest -= 1
+        elsif pattern[idx] == ',' && nest > 0
+          str << "|"
+        else
+          str << pattern[idx]
+        end
+        idx += 1
+      end
     end
-
-    Slice(UInt8*).new(paths.pathv, paths.pathc.to_i32).each do |path|
-      yield String.new(path)
+    return Regex.new("^#{regex_pattern}$")
```

Comment:
Should probably use `\A` and `\z` instead of `^` and `$` respectively, just to be on the safe side.